src/style/query_geometry.js

```diff
@@ -217,27 +281,244 @@ export class QueryGeometry {
+    const result = findEdgeCrossingAntimeridian(polygon, tr, containsNp ? -1 : 1);
+
+    if (!result) {
+        return null;
+    }
+
+    // Start constructing the new polygon by resampling edges until the crossing edge
+    const {idx, t} = result;
+    let partA = idx > 1 ? resamplePolygon(polygon.slice(0, idx), tr) : [];
+    let partB = idx < polygon.length ? resamplePolygon(polygon.slice(idx), tr) : [];
+
+    partA = partA.map(p => new Point(wrap(p.x), p.y));
+    partB = partB.map(p => new Point(wrap(p.x), p.y));
+
+    // Resample first section of the ring (up to the edge that crosses the 0-line)
+    const resampled = [...partA];
+
+    if (resampled.length === 0) {
+        resampled.push(partB[partB.length - 1]);
+    }
+
+    // Find location of the crossing by interpolating mercator coordinates.
+    // This will produce slightly off result as the crossing edge is not actually
+    // linear on the globe.
+    const a = resampled[resampled.length - 1];
+    const b = partB.length === 0 ? partA[0] : partB[0];
+    const intersectionY = interpolate(a.y, b.y, t);
+
+    let mid;
+
+    if (containsNp) {
+        mid = [
+            new Point(0, intersectionY),
+            new Point(0, 0),
+            new Point(1, 0),
+            new Point(1, intersectionY)
+        ];
+    } else {
+        mid = [
+            new Point(1, intersectionY),
+            new Point(1, 1),
+            new Point(0, 1),
+            new Point(0, intersectionY)
+        ];
+    }
+
+    resampled.push(...mid);
+
+    // Resample to the second section of the ring
+    if (partB.length === 0) {
+        resampled.push(partA[0]);
+    } else {
+        resampled.push(...partB);
+    }
+
+    return {
+        polygon: resampled.map(p => new MercatorCoordinate(p.x, p.y)),
+        unwrapped: false
+    };
+}
+
+function resamplePolygon(polygon: Point[], transform: Transform): Point[] {
+    const tolerance = 1.0 / 256.0;
+    return resample(
+        polygon,
+        p => {
+            const mc = transform.pointCoordinate3D(p);
+            p.x = mc.x;
+            p.y = mc.y;
+        },
+        tolerance);
+}
+
+function wrap(mercatorX: number): number {
+    return mercatorX < 0 ? 1 + (mercatorX % 1) : mercatorX % 1;
+}
+
+function findEdgeCrossingAntimeridian(polygon: Point[], tr: Transform, direction: number): ?{idx: number, t: number} {
+    for (let i = 1; i < polygon.length; i++) {
+        const a = wrap(tr.pointCoordinate3D(polygon[i - 1]).x);
+        const b = wrap(tr.pointCoordinate3D(polygon[i]).x);
+
+        // direction < 0: mercator coordinate 0 will be crossed from left
+        // direction > 0: mercator coordinate 1 will be crossed from right
+        if (direction < 0) {
+            if (a < b) {
+                return {idx: i, t: -a / (b - 1 - a)};
+            }
+        } else {
+            if (b < a) {
+                return {idx: i, t: (1 - a) / (b + 1 - a)};
+            }
+        }
+    }
```

Comment:
We search for one intersection, but how are polygons crossing the antimeridian several times handled?